File: src/main/cpp/commands/DriveTrajectory.cpp
// Copyright (c) 2025 FRC 6328
// http://github.com/Mechanical-Advantage
//
// Use of this source code is governed by an MIT-style
// license that can be found in the LICENSE file at
// the root directory of this project.

#include "drivetrajectory.h"

#include <cmath>
#include <functional>
#include <optional>
#include <vector>

#include <frc/Timer.h>
#include <frc/controller/PIDController.h>
#include <frc/controller/ProfiledPIDController.h>
#include <frc/geometry/Pose2d.h>
#include <frc/geometry/Rotation2d.h>
#include <frc/geometry/Translation2d.h>
#include <frc/kinematics/ChassisSpeeds.h>
#include <frc/trajectory/TrapezoidProfile.h>
#include <frc/util/Units.h>

#include "constants.h"
#include "junction/logger.h"
#include "robotstate.h"
#include "subsystems/drive/drive.h"
#include "util/alliancefliputil.h"
#include "util/loggedtunablenumber.h"
#include "util/mirrorutil.h"

namespace org::littletonrobotics::frc2025::commands {

LoggedTunableNumber DriveTrajectory::linearkP{
    "DriveTrajectory/LinearkP",
    Constants::GetRobot() == Constants::Robot::COMPBOT ||
            Constants::GetRobot() == Constants::Robot::DEVBOT
        ? 8.0
        : 4.0};
LoggedTunableNumber DriveTrajectory::linearkD{"DriveTrajectory/LinearkD", 0.0};
LoggedTunableNumber DriveTrajectory::thetakP{
    "DriveTrajectory/ThetakP",
    Constants::GetRobot() == Constants::Robot::COMPBOT ||
            Constants::GetRobot() == Constants::Robot::DEVBOT
        ? 4.0
        : 12.0};
LoggedTunableNumber DriveTrajectory::thetakD{"DriveTrajectory/thetakD", 0.0};
LoggedTunableNumber DriveTrajectory::overrideMaxVelocity{
    "DriveTrajectory/OverrideMaxVelocity", frc::Units::DegreesToRadians(360)};
LoggedTunableNumber DriveTrajectory::overrideMaxAcceleration{
    "DriveTrajectory/OverrideMaxAcceleration",
    frc::Units::DegreesToRadians(720)};

DriveTrajectory::DriveTrajectory(
    Drive &drive, subsystems::drive::trajectory::HolonomicTrajectory trajectory)
    : DriveTrajectory{drive, trajectory, false} {}

DriveTrajectory::DriveTrajectory(
    Drive &drive, subsystems::drive::trajectory::HolonomicTrajectory trajectory,
    bool mirror)
    : DriveTrajectory{
          drive, trajectory,
          [&]() { return RobotState::GetInstance().GetEstimatedPose(); },
          mirror} {}

DriveTrajectory::DriveTrajectory(
    Drive &drive, subsystems::drive::trajectory::HolonomicTrajectory trajectory,
    std::function<frc::Pose2d()> robot, bool mirror)
    : drive_{drive}, trajectory_{trajectory}, robotPose_{robot},
      mirror_{mirror}, thetaController_{0.0, 0.0, 0.0},
      overrideThetaController_{0.0, 0.0, 0.0,
                               frc::TrapezoidProfile::Constraints{0.0, 0.0}} {
  thetaController_.EnableContinuousInput(-M_PI, M_PI);
  overrideThetaController_.EnableContinuousInput(-M_PI, M_PI);
  AddRequirements({&drive_});
}

void DriveTrajectory::Initialize() {
  timer_.Restart();

  xController_.Reset();
  yController_.Reset();
  thetaController_.Reset();

  std::vector<frc::Pose2d> trajectoryPoses;
  for (const auto &pose : trajectory_.GetTrajectoryPoses()) {
    trajectoryPoses.push_back(
        AllianceFlipUtil::Apply(mirror_ ? MirrorUtil::Apply(pose) : pose));
  }
  Logger::RecordOutput("Trajectory/TrajectoryPoses", trajectoryPoses);
}

void DriveTrajectory::Execute() {
  if (linearkP.HasChanged(this->GetHashCode()) ||
      linearkD.HasChanged(this->GetHashCode())) {
    xController_.SetPID(linearkP.Get(), 0.0, linearkD.Get());
    yController_.SetPID(linearkP.Get(), 0.0, linearkD.Get());
  }
  if (thetakP.HasChanged(this->GetHashCode()) ||
      thetakD.HasChanged(this->GetHashCode())) {
    thetaController_.SetPID(thetakP.Get(), 0.0, thetakD.Get());
    overrideThetaController_.SetPID(thetakP.Get(), 0.0, thetakD.Get());
  }
  if (overrideMaxVelocity.HasChanged(this->GetHashCode()) ||
      overrideMaxAcceleration.HasChanged(this->GetHashCode())) {
    overrideThetaController_.SetConstraints(frc::TrapezoidProfile::Constraints{
        overrideMaxVelocity.Get(), overrideMaxAcceleration.Get()});
  }

  frc::Pose2d robot = robotPose_();
  auto setpointState = AllianceFlipUtil::Apply(
      mirror_ ? MirrorUtil::Apply(trajectory_.Sample(timer_.Get()))
              : trajectory_.Sample(timer_.Get()));

  double xFeedback = xController_.Calculate(robot.X(), setpointState.x());
  double yFeedback = yController_.Calculate(robot.Y(), setpointState.y());
  double thetaFeedback = thetaController_.Calculate(robot.Rotation().Radians(),
                                                    setpointState.theta());

  std::vector<frc::Translation2d::Vector> moduleForces;
  for (const auto &forces : setpointState.module_forces()) {
    moduleForces.emplace_back(
        frc::Translation2d{forces.fx(), forces.fy()}
            .RotateBy(frc::Rotation2d{setpointState.theta()}.UnaryMinus())
            .ToVector());
  }

  double finalOmega =
      overrideRotation_.has_value()
          ? overrideThetaController_.Calculate(
                robot.Rotation().Radians(), overrideRotation_.value().Radians())
          : thetaFeedback + setpointState.omega();

  drive_.RunVelocity(frc::ChassisSpeeds::FromFieldRelativeSpeeds(
                         xFeedback + setpointState.vx(),
                         yFeedback + setpointState.vy(), finalOmega,
                         frc::Rotation2d::Degrees(robot.Rotation().Degrees())),
                     moduleForces);

  Logger::RecordOutput("Trajectory/RobotPose", robot);
  Logger::RecordOutput("Trajectory/SetpointPose", setpointState.pose());
  Logger::RecordOutput(
      "Trajectory/Feedback",
      frc::Pose2d{xFeedback, yFeedback, frc::Rotation2d{thetaFeedback}});
  Logger::RecordOutput("Trajectory/VelocityFeedforward",
                       frc::Pose2d{setpointState.vx(), setpointState.vy(),
                                   frc::Rotation2d{setpointState.omega()}});
  Logger::RecordOutput(
      "Trajectory/OverrideRotation",
      overrideRotation_.has_value()
          ? frc::Pose2d{setpointState.pose().Translation(),
                        frc::Rotation2d{
                            overrideThetaController_.GetSetpoint().position}}
          : frc::Pose2d{});
}

void DriveTrajectory::End(bool interrupted) {}

bool DriveTrajectory::IsFinished() {
  return timer_.Get() >= trajectory_.GetDuration();
}

void DriveTrajectory::SetOverrideRotation(
    std::optional<frc::Rotation2d> rotation) {
  overrideRotation_ = rotation;
  if (!isOverrideRotation_.get()) {
    overrideThetaController_.Reset(
        robotPose_().Rotation().Radians(),
        RobotState::GetInstance().GetRobotVelocity().omega);
  }
  isOverrideRotation_.set(rotation.has_value());
}

} // namespace org::littletonrobotics::frc2025::commands

File: src/main/java/org/littletonrobotics/frc2025/commands/DriveTrajectory.java
// Copyright (c) 2025 FRC 6328
// http://github.com/Mechanical-Advantage
//
// Use of this source code is governed by an MIT-style
// license that can be found in the LICENSE file at
// the root directory of this project.

package org.littletonrobotics.frc2025.commands;

import static org.littletonrobotics.vehicletrajectoryservice.VehicleTrajectoryServiceOuterClass.*;

import edu.wpi.first.math.controller.PIDController;
import edu.wpi.first.math.controller.ProfiledPIDController;
import edu.wpi.first.math.geometry.Pose2d;
import edu.wpi.first.math.geometry.Rotation2d;
import edu.wpi.first.math.geometry.Translation2d;
import edu.wpi.first.math.kinematics.ChassisSpeeds;
import edu.wpi.first.math.trajectory.TrapezoidProfile;
import edu.wpi.first.math.util.Units;
import edu.wpi.first.wpilibj.Timer;
import edu.wpi.first.wpilibj2.command.Command;
import java.util.Arrays;
import java.util.Optional;
import java.util.function.Supplier;
import lombok.experimental.ExtensionMethod;
import org.littletonrobotics.frc2025.Constants;
import org.littletonrobotics.frc2025.RobotState;
import org.littletonrobotics.frc2025.subsystems.drive.Drive;
import org.littletonrobotics.frc2025.subsystems.drive.trajectory.HolonomicTrajectory;
import org.littletonrobotics.frc2025.subsystems.drive.trajectory.TrajectoryGenerationHelpers;
import org.littletonrobotics.frc2025.util.AllianceFlipUtil;
import org.littletonrobotics.frc2025.util.LoggedTunableNumber;
import org.littletonrobotics.frc2025.util.MirrorUtil;
import org.littletonrobotics.junction.AutoLogOutput;
import org.littletonrobotics.junction.Logger;

@ExtensionMethod({TrajectoryGenerationHelpers.class})
public class DriveTrajectory extends Command {
  private static final LoggedTunableNumber linearkP =
      new LoggedTunableNumber("DriveTrajectory/LinearkP");
  private static final LoggedTunableNumber linearkD =
      new LoggedTunableNumber("DriveTrajectory/LinearkD");
  private static final LoggedTunableNumber thetakP =
      new LoggedTunableNumber("DriveTrajectory/ThetakP");
  private static final LoggedTunableNumber thetakD =
      new LoggedTunableNumber("DriveTrajectory/ThetakD");
  private static final LoggedTunableNumber overrideMaxVelocity =
      new LoggedTunableNumber("DriveTrajectory/OverrideMaxVelocity", Units.degreesToRadians(360));
  private static final LoggedTunableNumber overrideMaxAcceleration =
      new LoggedTunableNumber(
          "DriveTrajectory/OverrideMaxAcceleration", Units.degreesToRadians(720));

  static {
    switch (Constants.getRobot()) {
      case COMPBOT, DEVBOT -> {
        linearkP.initDefault(8.0);
        linearkD.initDefault(0.0);
        thetakP.initDefault(4.0);
        thetakD.initDefault(0.0);
      }
      default -> {
        linearkP.initDefault(4.0);
        linearkD.initDefault(0.0);
        thetakP.initDefault(12.0);
        thetakD.initDefault(0.0);
      }
    }
  }

  private final Drive drive;
  private final HolonomicTrajectory trajectory;
  private final Timer timer = new Timer();
  private final Supplier<Pose2d> robotPose;
  private final boolean mirror;

  private Optional<Rotation2d> overrideRotation = Optional.empty();
  @AutoLogOutput private boolean isOverrideRotation = false;

  private final PIDController xController = new PIDController(0.0, 0.0, 0.0);
  private final PIDController yController = new PIDController(0.0, 0.0, 0.0);
  private final PIDController thetaController = new PIDController(0.0, 0.0, 0.0);
  private final ProfiledPIDController overrideThetaController =
      new ProfiledPIDController(0.0, 0.0, 0.0, new TrapezoidProfile.Constraints(0.0, 0.0));

  public DriveTrajectory(Drive drive, HolonomicTrajectory trajectory) {
    this(drive, trajectory, false);
  }

  public DriveTrajectory(Drive drive, HolonomicTrajectory trajectory, boolean mirror) {
    this(drive, trajectory, () -> RobotState.getInstance().getEstimatedPose(), mirror);
  }

  public DriveTrajectory(
      Drive drive, HolonomicTrajectory trajectory, Supplier<Pose2d> robot, boolean mirror) {
    this.drive = drive;
    this.trajectory = trajectory;
    robotPose = robot;
    this.mirror = mirror;

    thetaController.enableContinuousInput(-Math.PI, Math.PI);
    overrideThetaController.enableContinuousInput(-Math.PI, Math.PI);
    addRequirements(drive);
  }

  @Override
  public void initialize() {
    timer.restart();

    xController.reset();
    yController.reset();
    thetaController.reset();

    Logger.recordOutput(
        "Trajectory/TrajectoryPoses",
        Arrays.stream(trajectory.getTrajectoryPoses())
            .map(pose -> AllianceFlipUtil.apply(mirror ? MirrorUtil.apply(pose) : pose))
            .toArray(Pose2d[]::new));
  }

  @Override
  public void execute() {
    // Update tunable numbers
    if (linearkP.hasChanged(hashCode()) || linearkD.hasChanged(hashCode())) {
      xController.setPID(linearkP.get(), 0.0, linearkD.get());
      yController.setPID(linearkP.get(), 0.0, linearkD.get());
    }
    if (thetakP.hasChanged(hashCode()) || thetakD.hasChanged(hashCode())) {
      thetaController.setPID(thetakP.get(), 0.0, thetakD.get());
      overrideThetaController.setPID(thetakP.get(), 0.0, thetakD.get());
    }
    if (overrideMaxVelocity.hasChanged(hashCode())
        || overrideMaxAcceleration.hasChanged(hashCode())) {
      overrideThetaController.setConstraints(
          new TrapezoidProfile.Constraints(
              overrideMaxVelocity.get(), overrideMaxAcceleration.get()));
    }

    Pose2d robot = robotPose.get();
    // Get setpoint state and flip
    VehicleState setpointState =
        AllianceFlipUtil.apply(
            mirror
                ? MirrorUtil.apply(trajectory.sample(timer.get()))
                : trajectory.sample(timer.get()));

    // Calculate feedback
    double xFeedback = xController.calculate(robot.getX(), setpointState.getX());
    double yFeedback = yController.calculate(robot.getY(), setpointState.getY());
    double thetaFeedback =
        thetaController.calculate(robot.getRotation().getRadians(), setpointState.getTheta());

    // Calculate module forces
    var moduleForces =
        setpointState.getModuleForcesList().stream()
            .map(
                forces ->
                    new Translation2d(forces.getFx(), forces.getFy())
                        .rotateBy(Rotation2d.fromRadians(setpointState.getTheta()).unaryMinus())
                        .toVector())
            .toList();

    // Calculate final omega using override rotation if present
    final double finalOmega =
        overrideRotation
            .map(
                setpoint ->
                    overrideThetaController.calculate(
                        robot.getRotation().getRadians(), setpoint.getRadians()))
            .orElse(thetaFeedback + setpointState.getOmega());

    // Command drive
    drive.runVelocity(
        ChassisSpeeds.fromFieldRelativeSpeeds(
            xFeedback + setpointState.getVx(),
            yFeedback + setpointState.getVy(),
            finalOmega,
            Rotation2d.fromDegrees(robot.getRotation().getDegrees())),
        moduleForces);

    // Log outputs
    Logger.recordOutput("Trajectory/RobotPose", robot);
    Logger.recordOutput("Trajectory/SetpointPose", setpointState.getPose());
    Logger.recordOutput(
        "Trajectory/Feedback",
        new Pose2d(xFeedback, yFeedback, Rotation2d.fromRadians(thetaFeedback)));
    Logger.recordOutput(
        "Trajectory/VelocityFeedforward",
        new Pose2d(
            setpointState.getVx(),
            setpointState.getVy(),
            Rotation2d.fromRadians(setpointState.getOmega())));
    Logger.recordOutput(
        "Trajectory/OverrideRotation",
        overrideRotation
            .map(
                rotation ->
                    new Pose2d(
                        setpointState.getPose().getTranslation(),
                        Rotation2d.fromRadians(overrideThetaController.getSetpoint().position)))
            .orElse(Pose2d.kZero));
  }

  @Override
  public void end(boolean interrupted) {}

  @Override
  public boolean isFinished() {
    return timer.get() >= trajectory.getDuration();
  }

  public void setOverrideRotation(Optional<Rotation2d> rotation) {
    overrideRotation = rotation;
    if (!isOverrideRotation) {
      overrideThetaController.reset(
          robotPose.get().getRotation().getRadians(),
          RobotState.getInstance().getRobotVelocity().omegaRadiansPerSecond);
    }
    isOverrideRotation = rotation.isPresent();
  }
}
